src/controllers/orderController.js

```diff
@@ -5,58 +5,89 @@ const { BadRequestError } = require('../errors');
+
+  try {
+    session.startTransaction();
+
+    const { orderItems, subtotal, tax, total } = req.body;
+
+    if (!orderItems?.length) {
+      throw new BadRequestError(
+        'Unable to create an order because the order items are missing.'
+      );
+    }
+    // Fetch user information from the request object
+    const user = req.user; //contains user object
+    console.log(user);
+
+    const order = new Order({
+      user: req.user.userId,
+      orderItems, // Use the array with full album data
+      subtotal,
+      tax,
+      total,
+    });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: Math.round(total * 100), // Stripe requires amount in integer cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
+    await order.save({ session });
+
+    for (const orderItem of orderItems) {
+      const purchasedAlbum = new PurchasedAlbum({
+        album: orderItem.album,
+        user: req.user.userId,
+      });
+
+      await purchasedAlbum.save({ session });
+    }
+    //Server  tells Stripe that an order is being made (creating a "paymentIntent" object in Stripe)
+    const paymentIntent = await stripe.paymentIntents.create({
+      amount: total * 100,
+      currency: 'usd',
+      description: `Order #${order._id}`,
+      metadata: {
+        userId: req.user.userId,
+        orderId: order._id,
+        totalQuantity: orderItems.reduce(
+          (total, item) => total + item.quantity,
+          0
+        ),
+        totalAlbums: orderItems.length,
+      },
+    });
+    //Stripe  gives the server the paymentIntent object. The server will save the id of this object in the Order document so that we can easily associate the two.
+    order.paymentIntentId = paymentIntent.id;
+    await order.save({ session });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
+    await session.commitTransaction();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+    const orderItemsWithFullAlbum = await order.populate({
+      path: 'orderItems.album',
     });
+    // Send the order completion email after THE ORDER HAS STATUS COMPLETE
+    await sendOrderCompletedEmail(
+      user.email,
+      user.name,
+      orderItemsWithFullAlbum.orderItems,
+      orderItemsWithFullAlbum.total
+    );
```

Comment:
We can remove this from here because we don't want to send the email at this stage yet. At this point, the payment intent has just been submitted to Stripe; but the user hasn't actually put in the payment information and we don't know if the payment will be successful. So we likely don't want to send a "your order has completed" email. We _could_ have a separate email that says "we got your order request, please proceed to checkout" but that may get a bit too busy since the user could still abandon their order at this point of the process.

Reply:
@akosasante oh that totally makes sence!! thank you!